witness/src/Dscp/Snowdrop/PublicationValidation.hs

```diff
@@ -0,0 +1,106 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE DataKinds           #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+module Dscp.Snowdrop.PublicationValidation
+       ( validatePublication
+       , Publication (..)
+       , PublicationValidationError (..)
+       , PublicationTx (..)
+       , PublicationTxTypeId (..)
+       ) where
+
+import Data.Map as Map (lookup)
+
+import Dscp.Core.Transactions
+import Dscp.Snowdrop.AccountValidation
+import Snowdrop.Model.State.Core (HasKeyValue, PreValidator (..), StateTx (..), StateTxType (..),
+                                  Validator, mkValidator, queryOne)
+
+import Snowdrop.Util
+
+-- | Transaction type for publication.
+data PublicationTxTypeId
+    = PublicationTxTypeId
+    deriving (Eq, Ord, Show, Generic)
+
+data PublicationValidationError
+    = PublicationSignatureIsIncorrect
+    | PublicationPrevBlockIsIncorrect
+    | StorageIsCorrupted
+    | PublicationIsBroken
+
+validatePublication
+    :: forall       e pk addr id proof value ctx sig hash stTxId
+    .  ValidatorCtx e pk addr id proof value ctx sig hash stTxId (Publication hash addr)
+    => HasReview    e PublicationValidationError
+    => IdStorage                                          stTxId PublicationTxTypeId
+    => HasKeyValue  id value (PublicationsOf  addr) (LastPublication hash)
+    => HasKeyValue  id value (PublicationHead hash) (PublicationNext hash)
+    => Validator    e         id proof value ctx
+
+validatePublication = mkValidator ty
+    [preValidatePublication @e @pk @addr @id @proof @value @ctx @sig @hash @stTxId]
+  where
+    ty = StateTxType $ getId (Proxy @stTxId) PublicationTxTypeId
+
+preValidatePublication
+    :: forall       e pk addr id proof value ctx sig hash stTxId
+    .  ValidatorCtx e pk addr id proof value ctx sig hash stTxId (Publication hash addr)
+    => HasKeyValue  id value (PublicationsOf  addr) (LastPublication hash)
+    => HasKeyValue  id value (PublicationHead hash) (PublicationNext hash)
+    => HasReview    e PublicationValidationError
+    => PreValidator e         id proof value ctx
+preValidatePublication =
+    PreValidator $ \_trans@ StateTx {..} -> do
+
+        -- Retrieving our Publication payload from proof
+        (AccountId authorId, _, Publication {..}) <- authenticate
+            @e @pk @addr @id @proof @value @ctx @sig @hash @stTxId
+            @(Publication hash addr)
+            txProof
+
+        -- Checking that prevBlockHash matches the one from storage.
+        -- This can be written as:
+        --
+        -- res <- queryOne ...
+        -- validateUnless (getLastPub <$> res == pPrevBlockHash) Pub...
+        --
+        -- ... but this shows the intent more clear.
+        lastPub <- queryOne (PublicationsOf authorId) >>= \it -> do
+            () <- case it of
+                Just (LastPublication theLast) -> do
```

Comment:
Redunant do.

Reply:
If you insert a line into monadic expression you didn't prefixed with `do`, you can forget the `do` and get `called with too many parameters` error. This makes you to recompile and loose your line of thought.

Thats why i place `do` in one-line blocks.

Reply:
It doesn't look like a serious argument, sorry -- use `stack --file-watch` if you don't want to recompile. Also, there are many other mistakes you can make. We need `hlint` integration, it would definitely help us with these problems.